tests/ga/extension_actor.py

```diff
@@ -0,0 +1,465 @@
+    function. Specifically, the *.zip resource urls for an extension must end in "{ExtensionName}__{ExtensionVersion}".
+
+    Note that this function does not update xml for ext_conf or manifest, meaning that added extensions must already be present
+    within those files as provided by the firstActor's data source. This could probably be implemented with the
+    WireProtocolDataBase.replace_xml_element_value function, but it would probably make sense to first implement autogeneration
+    of those files within ExtensionActor.__init__, as we would then know exactly the xml elements in those files. This would
+    simplify the logic of adding the new actors, but is a bigger change. Ultimately, this feature isn't currently (i.e. as of
+    writing) needed.
+    """
+
+    new_actors = list(filter(lambda actor: actor not in protocol._actors, actors))  # list() applies the lambda before we edit protocol._actors
+    protocol._actors.extend(new_actors)
+
+    for actor in new_actors:
+        patched_popen = actor.patch_popen() # Like in get_protocol_and_handler, the patch_popen() and start() calls must
+        patched_popen.start()               # be interleaved.
+
+        # Keep track of the patches so the finally clause within get_protocol_and_handler can stop them.
+        protocol._patched_popens.append(patched_popen)
+    
+    protocol.set_http_handlers(http_get_handler=_generate_mock_http_get(protocol._actors[1:]),
+        http_put_handler=_generate_mock_http_put(protocol._actors))
+
+    distinct_names = set(actor.extension_info.name for actor in protocol._actors)
+    for name in distinct_names:
+        
+        max_ver = max(map(lambda actor: FlexibleVersion(actor.extension_info.version),
+            filter(lambda actor: actor.extension_info.name == name, protocol._actors)))
+
+        protocol.mock_wire_data.set_specific_extension_config_version(name, max_ver)
+    
+    protocol.mock_wire_data.set_incarnation(incarnation)
+    protocol.client.update_goal_state()
+
+
+class ActionSet(object):
+    """
+    A wrapper class for the possible actions that an extension much support.
+    """
+
+    def __init__(self, installAction, uninstallAction, updateAction, enableAction, disableAction, name, version):
+
+        # The keys below will be used to configure attributes for a mock; periods in such attributes are treated
+        # as attribute trees, which would break introspection.
+        formattedId = Formats.FormatExtensionDir(name, version).replace(".", "_")
+
+        self.delegate ={
+            "installCommand": dict(key="{0}_install".format(formattedId), action=installAction),
+            "uninstallCommand": dict(key="{0}_uninstall".format(formattedId), action=uninstallAction),
+            "updateCommand": dict(key="{0}_update".format(formattedId), action=updateAction),
+            "enableCommand": dict(key="{0}_enable".format(formattedId), action=Actions._wrap_with_write_status(enableAction, name, version)),
```

Comment:
All extension commands (not only enable) report status

Reply:
The agent only seems to check the status file after enable is called. Is this behavior expected?

Reply:
The current implementation checks for status at the end of the loop, but this should actually be async to extension execution. For example, extensions may be reporting progress of their install, and the agent should report that status to users.

Even in today's implementation, if install succeeds and reports status and then enable crashes before reporting status, the agent would report the status from install.

And lastly, the REST API allows users to disable extensions. The disable command should also report status and the agent should pick it up and report it up the chain.